File: internal/manager/agent/files.go
package agent

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"time"
)

type backupStage string

const (
	backupStagePermission backupStage = "permission"
	backupStageWrite      backupStage = "write"
	backupStageSync       backupStage = "sync"
	backupStageReopen     backupStage = "reopen"
	backupStageRead       backupStage = "read"
	backupStageIdentity   backupStage = "identity"
	backupStageContent    backupStage = "content"
)

type postReplaceError struct{ err error }

func (e *postReplaceError) Error() string { return e.err.Error() }
func (e *postReplaceError) Unwrap() error { return e.err }

func replacementOccurred(err error) bool {
	var replaced *postReplaceError
	return errors.As(err, &replaced)
}

func createPrivateBackup(sourcePath string, content []byte, sourceMode os.FileMode, label string) (string, error) {
	return createPrivateBackupWithHook(sourcePath, content, sourceMode, label, nil)
}

func createPrivateBackupWithHook(sourcePath string, content []byte, sourceMode os.FileMode, label string, hook func(backupStage, string) error) (string, error) {
	dir := filepath.Dir(sourcePath)
	base := filepath.Base(sourcePath)
	for attempts := 0; attempts < 16; attempts++ {
		var random [12]byte
		if _, err := io.ReadFull(rand.Reader, random[:]); err != nil {
			return "", err
		}
		name := fmt.Sprintf("%s.%s-%s-%s", base, label, time.Now().UTC().Format("20060102-150405.000000000"), hex.EncodeToString(random[:]))
		path := filepath.Join(dir, name)
		file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
		if errors.Is(err, os.ErrExist) {
			continue
		}
		if err != nil {
			return "", err
		}
		ok := false
		cleanup := func() error {
			file.Close()
			if !ok {
				return removeAndSync(path)
			}
			return nil
		}
		fail := func(cause error) (string, error) {
			if cleanupErr := cleanup(); cleanupErr != nil {
				return path, errors.Join(cause, cleanupErr)
			}
			return "", cause
		}
		if err := runBackupHook(hook, backupStagePermission, path); err != nil {
			return fail(err)
		}
		if err := restrictPrivate(path, false); err != nil {
			return fail(err)
		}
		if err := runBackupHook(hook, backupStageWrite, path); err != nil {
			return fail(err)
		}
		if _, err := file.Write(content); err != nil {
			return fail(err)
		}
		if err := runBackupHook(hook, backupStageSync, path); err != nil {
			return fail(err)
		}
		if err := file.Sync(); err != nil {
			return fail(err)
		}
		if err := file.Close(); err != nil {
			return fail(err)
		}
		if err := applyPrivateMode(path, sourceMode); err != nil {
			return fail(err)
		}
		if err := verifyPrivateBackup(path, content, hook); err != nil {
			return fail(err)
		}
		if err := syncDirectory(dir); err != nil {
			return fail(err)
		}
		ok = true
		_ = cleanup()
		return path, nil
	}
	return "", errors.New("could not allocate unique backup path")
}

func verifyPrivateBackup(path string, expected []byte, hook func(backupStage, string) error) error {
	before, err := os.Lstat(path)
	if err != nil || isFinalComponentLink(path, before) || !before.Mode().IsRegular() {
		return errors.New("backup identity is unsafe")
	}
	if err := runBackupHook(hook, backupStageReopen, path); err != nil {
		return err
	}
	file, err := os.Open(path)
	if err != nil {
		return err
	}
	defer file.Close()
	after, err := file.Stat()
	if err != nil || !after.Mode().IsRegular() || !os.SameFile(before, after) {
		return errors.New("backup identity changed")
	}
	if err := runBackupHook(hook, backupStageIdentity, path); err != nil {
		return err
	}
	pathAfter, err := os.Lstat(path)
	if err != nil || isFinalComponentLink(path, pathAfter) || !pathAfter.Mode().IsRegular() || !os.SameFile(after, pathAfter) {
		return errors.New("backup path identity changed")
	}
	if err := runBackupHook(hook, backupStageRead, path); err != nil {
		return err
	}
	content, err := io.ReadAll(io.LimitReader(file, maxConfigSize+1))
	if err != nil || len(content) > maxConfigSize {
		return errors.New("backup read-back failed")
	}
	if err := runBackupHook(hook, backupStageContent, path); err != nil {
		return err
	}
	if !bytes.Equal(content, expected) {
		return errors.New("backup content mismatch")
	}
	return nil
}

func runBackupHook(hook func(backupStage, string) error, stage backupStage, path string) error {
	if hook == nil {
		return nil
	}
	return hook(stage, path)
}

func writeAtomic(path string, content []byte, mode os.FileMode, private bool) (err error) {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	tmp, err := os.CreateTemp(dir, "."+filepath.Base(path)+"-*")
	if err != nil {
		return err
	}
	tmpPath := tmp.Name()
	defer func() {
		tmp.Close()
		os.Remove(tmpPath)
	}()
	// Keep secret-bearing temporary contents user-private until replacement.
	if err := restrictPrivate(tmpPath, false); err != nil {
		return err
	}
	if _, err := tmp.Write(content); err != nil {
		return err
	}
	if err := tmp.Sync(); err != nil {
		return err
	}
	if err := tmp.Close(); err != nil {
		return err
	}
	if !private {
		if err := applyTargetPermissions(tmpPath, path, mode); err != nil {
			return err
		}
	}
	if err := replaceAtomic(tmpPath, path); err != nil {
		return err
	}
	if err := syncDirectory(dir); err != nil {
		return &postReplaceError{err: err}
	}
	return nil
}

File: desktop/src-tauri/src/manager_core/agent/files.rs
use std::fs::{self, File, OpenOptions};
use std::io::{self, ErrorKind, Write};
use std::path::{Path, PathBuf};

use uuid::Uuid;

use super::detect::read_config;
use super::recovery::is_final_component_link;
use super::types::MAX_CONFIG_SIZE;

#[cfg(not(windows))]
use super::files_unix as platform;
#[cfg(windows)]
use super::files_windows as platform;

pub use platform::{
    apply_private_mode, apply_target_permissions, private_permissions_ok, replace_atomic,
    restrict_private, sync_directory,
};

/// Accept an already-private object, or tighten a current-user inherited ACL
/// left by older Windows builds whose restrict helper was a no-op.
pub fn ensure_private_permissions(path: &Path, directory: bool, mode: u32) -> bool {
    if private_permissions_ok(path, directory, mode) {
        return true;
    }
    if !cfg!(windows) {
        return false;
    }
    restrict_private(path, directory).is_ok() && private_permissions_ok(path, directory, mode)
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum BackupStage {
    Permission,
    Write,
    Sync,
    Reopen,
    Read,
    Identity,
    Content,
}

#[derive(Debug)]
pub struct PostReplaceError {
    pub source: io::Error,
}

impl std::fmt::Display for PostReplaceError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.source.fmt(f)
    }
}

impl std::error::Error for PostReplaceError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        Some(&self.source)
    }
}

pub fn replacement_occurred(error: &(dyn std::error::Error + 'static)) -> bool {
    error.downcast_ref::<PostReplaceError>().is_some()
}

pub fn write_atomic(
    path: &Path,
    content: &[u8],
    mode: u32,
    private: bool,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let dir = path
        .parent()
        .ok_or_else(|| io::Error::new(ErrorKind::InvalidInput, "missing parent"))?;
    fs::create_dir_all(dir)?;
    let tmp_path = dir.join(format!(
        ".{}-{}",
        path.file_name()
            .and_then(|value| value.to_str())
            .unwrap_or("tmp"),
        Uuid::new_v4().simple()
    ));
    let written = (|| {
        {
            let mut tmp = OpenOptions::new()
                .write(true)
                .create_new(true)
                .open(&tmp_path)?;
            restrict_private(&tmp_path, false)?;
            tmp.write_all(content)?;
            tmp.sync_all()?;
        }
        if !private {
            apply_target_permissions(&tmp_path, path, mode)?;
        }
        replace_atomic(&tmp_path, path)?;
        Ok::<(), Box<dyn std::error::Error + Send + Sync>>(())
    })();
    if written.is_err() {
        let _ = fs::remove_file(&tmp_path);
    }
    written?;
    if let Err(error) = sync_directory(dir) {
        return Err(Box::new(PostReplaceError { source: error }));
    }
    Ok(())
}

pub fn remove_target_and_sync(path: &Path) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    fs::remove_file(path)?;
    if let Err(error) = sync_directory(path.parent().unwrap_or(path)) {
        return Err(Box::new(PostReplaceError { source: error }));
    }
    Ok(())
}

pub fn remove_and_sync(path: &Path) -> io::Result<()> {
    if let Err(error) = fs::remove_file(path) {
        if error.kind() != ErrorKind::NotFound {
            return Err(error);
        }
    }
    sync_directory(path.parent().unwrap_or(path))
}

pub fn create_private_backup(
    source_path: &Path,
    content: &[u8],
    source_mode: u32,
    label: &str,
) -> Result<PathBuf, Box<dyn std::error::Error + Send + Sync>> {
    create_private_backup_with_hook(source_path, content, source_mode, label, None)
}

pub fn create_private_backup_with_hook(
    source_path: &Path,
    content: &[u8],
    source_mode: u32,
    label: &str,
    hook: Option<
        &dyn Fn(BackupStage, &Path) -> Result<(), Box<dyn std::error::Error + Send + Sync>>,
    >,
) -> Result<PathBuf, Box<dyn std::error::Error + Send + Sync>> {
    let dir = source_path
        .parent()
        .ok_or_else(|| io::Error::new(ErrorKind::InvalidInput, "missing parent"))?;
    let base = source_path
        .file_name()
        .and_then(|value| value.to_str())
        .unwrap_or("backup");
    for _ in 0..16 {
        let name = format!(
            "{base}.{label}-{}-{}",
            chrono::Utc::now().format("%Y%m%d-%H%M%S.%f"),
            Uuid::new_v4().simple()
        );
        let path = dir.join(name);
        match OpenOptions::new().write(true).create_new(true).open(&path) {
            Ok(_) => {}
            Err(error) if error.kind() == ErrorKind::AlreadyExists => continue,
            Err(error) => return Err(error.into()),
        }
        if let Err(error) = finish_backup(&path, dir, content, source_mode, hook) {
            let _ = remove_and_sync(&path);
            return Err(error);
        }
        return Ok(path);
    }
    Err("could not allocate unique backup path".into())
}

fn finish_backup(
    path: &Path,
    dir: &Path,
    content: &[u8],
    source_mode: u32,
    hook: Option<
        &dyn Fn(BackupStage, &Path) -> Result<(), Box<dyn std::error::Error + Send + Sync>>,
    >,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    if let Some(hook) = hook {
        hook(BackupStage::Permission, path)?;
    }
    restrict_private(path, false)?;
    if let Some(hook) = hook {
        hook(BackupStage::Write, path)?;
    }
    {
        let mut file = OpenOptions::new().write(true).open(path)?;
        file.write_all(content)?;
        if let Some(hook) = hook {
            hook(BackupStage::Sync, path)?;
        }
        file.sync_all()?;
    }
    apply_private_mode(path, source_mode)?;
    verify_private_backup(path, content, hook)?;
    sync_directory(dir)?;
    Ok(())
}

fn verify_private_backup(
    path: &Path,
    expected: &[u8],
    hook: Option<
        &dyn Fn(BackupStage, &Path) -> Result<(), Box<dyn std::error::Error + Send + Sync>>,
    >,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let before = fs::symlink_metadata(path)?;
    if is_final_component_link(path, &before) || !before.is_file() {
        return Err("backup identity is unsafe".into());
    }
    if let Some(hook) = hook {
        hook(BackupStage::Reopen, path)?;
    }
    let file = File::open(path)?;
    let after = file.metadata()?;
    if !after.is_file() || !same_file(&before, &after) {
        return Err("backup identity changed".into());
    }
    if let Some(hook) = hook {
        hook(BackupStage::Identity, path)?;
    }
    let path_after = fs::symlink_metadata(path)?;
    if is_final_component_link(path, &path_after)
        || !path_after.is_file()
        || !same_file(&after, &path_after)
    {
        return Err("backup path identity changed".into());
    }
    if let Some(hook) = hook {
        hook(BackupStage::Read, path)?;
    }
    let content = read_config(path).map_err(|_| "backup read-back failed")?;
    if content.len() > MAX_CONFIG_SIZE {
        return Err("backup read-back failed".into());
    }
    if let Some(hook) = hook {
        hook(BackupStage::Content, path)?;
    }
    if content.as_slice() != expected {
        return Err("backup content mismatch".into());
    }
    Ok(())
}

fn same_file(left: &fs::Metadata, right: &fs::Metadata) -> bool {
    #[cfg(unix)]
    {
        use std::os::unix::fs::MetadataExt;
        left.dev() == right.dev() && left.ino() == right.ino()
    }
    #[cfg(not(unix))]
    {
        left.len() == right.len() && left.modified().ok() == right.modified().ok()
    }
}
